### DISCUSSION TITLE
Bug-Report? Multi-Shader-Setup Rendering

### DISCUSSION THREAD

**Discussion by dapa5900** [JacquesLucke/animation_nodes#1750] [General]
Hi @OmarEmaraDev 

I tried to create an Animation Nodes driven Multi-Shader-Setup for a Book generator I'm working on. I used an image sequence texture for that and drive the frame offset via the AN Material Attribute Output Node for the instanced materials.

It seems to work as intended rendering with Cycles in the viewport but when I want to render an image, Blender crashes every time. With Eevee it doesn't seem to work at all. I get some weird ghosting and sometimes the selected pages change. Sometimes I get the purple missing texture color and for the most part I can only see one of the page textures on all pages.

You can check the file here:

https://p7s1-my.sharepoint.com/:f:/g/personal/par0001d_belgium_fhm_de/ErguGp35s4tDuK-G2jeb8tcBcGrOXt0tPq9Iv2SVw_tzfQ?e=H4byEi

It's a bit messy but I highlighted all the material related nodes with a blue color. I actually packed all the textures into the file but I'm not sure if that worked correctly, as the blend-file size doesn't differ from the non-packed version so I also put the texture folder next to the project. 

When you have time, it would be great if you could check this out. Thank you! :-)

**Comment by 3DSinghVFX** [comment]
@dapa5900 Hi. In the *Material Instancer* node you need to keep off the option *Instance Material* after instancing the material. You may also look into this answer on BSE: https://blender.stackexchange.com/questions/153377/in-animation-nodes-is-it-possible-to-make-a-node-to-translate-position-of-the-u/153499#153499

**Reply by dapa5900** [reply]
@3DSinghVFX 

alright thanks a lot for your quick reply; this solved the problem for Cycles (I had to save and reload the file though otherwise it would still crash upon rendering). 

But for Eevee this workflow is not applicable/working? Is there maybe another method?

I think in general it would be great if via AN or even better Blender directly we could get a proper Multi-Texture workflow as I feel that this approach I used here is a pretty hacky way of doing this; also considering the development with GeoNodes right now. Maybe this would be something for you @OmarEmaraDev ...? If I remember correctly you once mentioned that you're working on Cycles as well... Maybe this could be implemented some day? If this sounds doable and interesting to you, please let me know, if I can help with a proper proposal for this. Thanks again!

**Reply by OmarEmaraDev** [reply]
I do agree that material instancing is not a good way do handle this. Do you know of any software that implements Multi-Texture workflow? What would something like that look like?

**Reply by dapa5900** [reply]
@OmarEmaraDev 

thank you for your reply and interest in this! :-) yes, you can definitely do this in C4Ds internal render and I also did this with Redshift before. It's been a while since I used both of them and since I'm now only using Blender I also unfortunately don't have access to both of them anymore but i will look things up and come up with a design proposal next week. Please let me know, wheres a good place to post this then. Thanks and have a nice weekend!

**Reply by OmarEmaraDev** [reply]
@dapa5900 Sounds good. Perhaps bounce your ideas off me first before writing a detailed proposal, I think I can provide some valuable feedback from an implementation point of view.

**Reply by dapa5900** [reply]
Hi @OmarEmaraDev ,

so, here is a short tutorial, showing, how the MultiShader works in C4D Standard Render:

https://www.youtube.com/watch?v=cFxw3EkBAAM&list=WL&index=5

And in this image you can see how the node looks in Redshift (had to add the folder comment because I could only get this cropped version from someone on Facebook):

![redshiftMultiShader](https://user-images.githubusercontent.com/75205162/124430540-96709100-dd6f-11eb-84b3-54798b587299.jpg)

So I think in it's most basic version it should be possible to auto-populate the contents of the node with textures, that you can batch-select from a folder and then output the color information through a single output socket. You would then need some kind of input that let's you define how the items will be distributed across the instances. So you could either have for example a drop down that let's you select between different modes (eg. sequentially, randomly with a seed value, or any other attribute input that can be controlled via GeoNodes and AN) or you just have an input socket that can accept the information from an Attribute node directly. 
As a bonus i think it would be good, if you can manually reorder items and add and delete individual input channels of the node (it would also be good to have an option to clear all items, in case you have quite a few). Finally I think it would be the best if every item also has as an input-socket, so that you can also plug in all kinds of color or procedural texture information from the shader graph for example.

Let me know, what you think, or if you have questions! :-)

**Reply by OmarEmaraDev** [reply]
@dapa5900 I see. I feel like there is a possibility we may be able to create an add-on that does this efficiently and transparently. I will have to give it a try, but I think it is possible. I will get back to you on this one.

**Reply by dapa5900** [reply]
@OmarEmaraDev 

great thanks!

btw: how's it going along with the looping noises from ticket #1639 ?

**Reply by OmarEmaraDev** [reply]
@dapa5900 Not great unfortunately, writing a library from scratch takes time and effort and I couldn't allocate either the last couple of months. I wouldn't promise a delivery time at this point, but make sure to ping me again when you can.

**Reply by dapa5900** [reply]
@OmarEmaraDev alright will do!

**Reply by OmarEmaraDev** [reply]
@dapa5900  What do you think about something like this?

https://user-images.githubusercontent.com/25300994/124659737-92df2600-dea5-11eb-948a-d658d74a9e5e.mp4

**Reply by dapa5900** [reply]
@OmarEmaraDev 

Thank you for setting this up so quickly, it already looks really fantastic.

Here's an image with my feedback so far on this:
 
![feedbackImageSelector](https://user-images.githubusercontent.com/75205162/124713686-26921000-df01-11eb-8f52-25ed6be567e0.jpg)

As said, the option to wire in eg. RGB-values (for example if you want to crate a color palette) or procedural noises from the shader graph via sockets would be nice to have and could maybe also implemented later on if possible. But I think for now this looks awesome and should be sufficient the implement the basic functionality how to assign these values to instances then for example.

**Reply by OmarEmaraDev** [reply]
@dapa5900 The circle does nothing, I just needed a way to indicate which image is selected, not sure if there is a better way to indicate that. Clicking the dot will just select the image. The Remove button doesn't remove from bottom to top, it removes the selected item indicated by the dot.

For having RGB sockets, I am not sure how feasible this is yet. I will look into it later. I will try to publish a working prototype soon so that you can test it.

**Reply by dapa5900** [reply]
@great! Looking forward to test it.

**Reply by OmarEmaraDev** [reply]
@dapa5900 Here is an initial version for testing, feel free to open an issue there for feedback.

https://github.com/OmarEmaraDev/shader-selector-nodes
